cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -298,11 +299,27 @@ private List<Runnable> processNotification(ProgramHeartbeatTable programHeartbea
       return result;
     }
 
+    //At this time, it is expected that provisioning service should be up and running.
+    waitForProvisioningService(provisioningService);
+
     handleClusterEvent(programRunId, clusterStatus, notification,
                        messageIdBytes, appMetadataStore, context).ifPresent(result::add);
     return result;
   }
 
+  //Waits maximum for 100 seconds.
+  private void waitForProvisioningService(ProvisioningService provisioningService) throws ServiceNotEnabledException {
+    RetryStrategy retryStrategy =
+      RetryStrategies.limit(100, RetryStrategies.fixDelay(1000, TimeUnit.MILLISECONDS));
+
+    //Wait Until Provisioning Service is running
+    Retries.runWithRetries(() -> {
+        if (!provisioningService.isRunning()) {
+          throw new ServiceNotEnabledException(provisioningService.getClass().getSimpleName());
```

Comment:
Assuming that this logic is going to be called during startup time, ServiceUnavailableException seems more appropriate.